### PR TITLE
OLS-419: More detailed errors in HTTP REST API responses

### PR DESCRIPTION
**Pull request by tisnik**
## Description

[OLS-419](https://issues.redhat.com//browse/OLS-419): More detailed errors in HTTP REST API responses

An example of error response (formatted using `jq`):

```
Sorry, an error occurred:

{
  "detail": {
    "response": "Error while validating question",
    "cause": "Failed to handle request to [https://bam-api.res.ibm.com/v2/text/generation/limits?version=2023-11-22.\n{\n](https://bam-api.res.ibm.com/v2/text/generation/limits?version=2023-11-22.%5Cn%7B%5Cn) 'error': 'Unauthorized',\n 'extensions': {\n 'code': 'AUTH_ERROR',\n 'state': null,\n 'reason': 'TOKEN_INVALID'\n },\n 'message': 'Invalid or missing JWT token',\n 'status_code': 401\n}"
  }
}
```

## Type of change

- [x] Refactor
- [ ] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Configuration Update
- [ ] Bump-up dependent library

## Related Tickets & Documents

- Related Issue #[OLS-419](https://issues.redhat.com//browse/OLS-419)
- Closes #

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] PR has passed all pre-merge test jobs.
- [ ] If it is a core feature, I have added thorough tests.

### CODE REVIEW

File: docs/openapi.json [hunk 1]
```diff
@@ -26,7 +26,7 @@
                 },
                 "responses": {
                     "200": {
-                        "description": "Successful Response",
+                        "description": "Query was validated and correct response from LLM is returned",
```
Comment (raptorsun) [inline docs/openapi.json:29]: how about  `s/was validated/is valid/`?

File: docs/openapi.json [hunk 2]
```diff
@@ -35,6 +35,16 @@
...
+                        "content": {
+                            "application/json": {
+                                "schema": {
+                                    "$ref": "#/components/schemas/ErrorResponse"
+                                }
+                            }
+                        }
+                    },
                     "422": {
                         "description": "Validation Error",
```
Comment (raptorsun) [inline docs/openapi.json:49]: Does this mean the query content is invalid but its format is good?

File: ols/app/endpoints/ols.py [hunk 3]
```diff
@@ -125,17 +137,27 @@ def retrieve_conversation_id(llm_request: LLMRequest) -> str:
...
+            logger.info(
+                f"{llm_request.conversation_id} Previous conversation input: {previous_input}"
+            )
+        return previous_input
+    except Exception as e:
+        logger.error(f"Error retrieving previous user input for user {user_id}")
+        raise HTTPException(
+            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+            detail={
+                "response": "Error retrieving previous user input",
```
Comment (raptorsun) [inline ols/app/endpoints/ols.py:157]: it seems less confusing "Error retrieving conversation history"

File: docs/openapi.json (hunk 2 above)
Comment (tisnik) [reply docs/openapi.json:49]: query content from LLM point of view. It's user query, not anything related to REST API, if your question is about it

File: ols/app/endpoints/ols.py (hunk 3 above)
Comment (tisnik) [reply ols/app/endpoints/ols.py:157]: indeed

Comment (tisnik) [reply ols/app/endpoints/ols.py]: updated

File: docs/openapi.json (hunk 1 above)
Comment (tisnik) [reply docs/openapi.json]: updated

File: ols/app/endpoints/ols.py [hunk 4]
```diff
@@ -26,7 +26,19 @@
...
-@router.post("/query")
+query_responses = {
+    200: {
+        "description": "Query is valid and correct response from LLM is returned",
+        "model": LLMResponse,
+    },
+    500: {
+        "description": "Query can not be validated, LLM is not accessible or other internal error",
+        "model": ErrorResponse,
+    },
```
Comment (bparees) [inline ols/app/endpoints/ols.py:37]: any reason not to also list 401 + 403 errors as potential response types?

Comment (tisnik) [reply ols/app/endpoints/ols.py:37]: we can, but I'm not sure about the model (i.e. about the actual format of generated error messages). Lemme check, I'll probably ask @nemerna

**Pull request merged**